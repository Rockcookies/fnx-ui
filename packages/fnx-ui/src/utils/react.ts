import { Children, ReactElement } from 'react';
import { isFragment } from 'react-is';

export const toElementArray = (
	children: React.ReactNode,
	option?: { keepEmpty?: boolean },
): ReactElement[] => {
	const keepEmpty = option && option.keepEmpty;
	let ret: React.ReactElement[] = [];

	Children.forEach(children, (child: any) => {
		if ((child === undefined || child === null) && !keepEmpty) {
			return;
		}

		if (Array.isArray(child)) {
			ret = ret.concat(toElementArray(child));
		} else if (isFragment(child) && child.props) {
			ret = ret.concat(toElementArray(child.props.children, option));
		} else {
			ret.push(child);
		}
	});

	return ret;
};
