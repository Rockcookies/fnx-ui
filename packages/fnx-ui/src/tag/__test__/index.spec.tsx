import React from 'react';
import { act, fireEvent, render } from '@testing-library/react';
import Tag from '../index';
import TestsDOM from '../../../test/dom';

describe('<Tag>', () => {
	beforeEach(() => {
		jest.useFakeTimers();
	});

	afterEach(() => {
		jest.useRealTimers();
	});
	it('should emit close event when clicking the close icon', () => {
		const onClose = jest.fn();

		const { container } = render(
			<Tag closeable onClose={onClose}>
				closeable
			</Tag>,
		);

		const track = TestsDOM.mustQuerySelector(container, '.fnx-tag__close');

		fireEvent.click(track);

		expect(onClose).toHaveBeenCalled();
	});

	it('should emit close event when clicking the close icon', () => {
		const onClose = jest.fn();

		const { container } = render(
			<Tag closeable onClose={onClose}>
				closeable
			</Tag>,
		);

		const track = TestsDOM.mustQuerySelector(container, '.fnx-tag__close');

		fireEvent.click(track);

		expect(onClose).toHaveBeenCalled();
	});

	it('when the theme is plain', () => {
		const { container } = render(<Tag theme="plain"></Tag>);

		expect(container).toMatchSnapshot();
	});
	it('transitionDuration', () => {
		const { container } = render(<Tag transitionDuration={100}></Tag>);

		expect(container).toMatchSnapshot();
	});

	it('can be controlled by visible with hidden as initial value', () => {
		const { container, rerender } = render(
			<Tag visible={true}>switch</Tag>,
		);

		expect(container).toMatchSnapshot();

		rerender(<Tag visible={false}>switch</Tag>);

		act(() => {
			jest.runAllTimers();
		});

		expect(container).toMatchSnapshot();

		rerender(<Tag visible={true}>switch</Tag>);

		act(() => {
			jest.runAllTimers();
		});

		expect(container).toMatchSnapshot();
	});
});
