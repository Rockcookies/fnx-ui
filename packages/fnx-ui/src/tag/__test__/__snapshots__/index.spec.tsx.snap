// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tag> can be controlled by visible with hidden as initial value 1`] = `
<div>
  <span
    class="fnx-tag fnx-tag--default"
  >
    switch
  </span>
</div>
`;

exports[`<Tag> can be controlled by visible with hidden as initial value 2`] = `<div />`;

exports[`<Tag> can be controlled by visible with hidden as initial value 3`] = `
<div>
  <span
    class="fnx-tag fnx-tag--default fnx-fade-enter-done"
  >
    switch
  </span>
</div>
`;

exports[`<Tag> transitionDuration 1`] = `
<div>
  <span
    class="fnx-tag fnx-tag--default"
    style="transition-duration: 100ms;"
  />
</div>
`;

exports[`<Tag> when the theme is plain 1`] = `
<div>
  <span
    class="fnx-tag fnx-tag--default fnx-tag--plain"
  />
</div>
`;
